src/seedu/addressbook/data/person/Contact.java

```diff
@@ -0,0 +1,60 @@
+package seedu.addressbook.data.person;
+
+import seedu.addressbook.data.exception.IllegalValueException;
+
+/**
+ * Represents a Person's attribute in the address book.
+ * Guarantees: immutable; is valid as declared in {@link #isValidValue(String)}
+ */
+public class Contact {
+
+    public final String value;
+    private boolean isPrivate;
+    
+    /**
+     * Validates given value.
+     *
+     * @throws IllegalValueException if given value string is invalid.
+     */
+    public Contact(String value, boolean isPrivate,
+            String validValueRegex, String invalidValueMessage) throws IllegalValueException {
+        String trimmedValue = value.trim();
+        if (!isValidValue(trimmedValue, validValueRegex)) {
+            throw new IllegalValueException(invalidValueMessage);
+        }
+        this.value = trimmedValue;
+        this.isPrivate = isPrivate;
+    }
+    
+    /**
+     * Returns true if a given string is a valid value against provided regex.
+     */
+    public static boolean isValidValue(String test, String validValueRegex) {
+        return test.matches(validValueRegex);
+    }
+    
+    public boolean isPrivate() {
+        return isPrivate;
+    }
+    
+    @Override
+    public int hashCode() {
+        return value.hashCode();
+    }
+    
+    @Override
+    public String toString() {
+        return value;
+    }
+    
+    /**
+     * Compares equality with another object. An object is considered equal if
+     * it is of the class Contact (or a subclass) and has the same value.
+     */
+    @Override
+    public boolean equals(Object other) {
```

Comment:
This is fine, but note that now an Email and Address having the same value will be considered equal even if they are of different types.

You can override at child classes but make sure to override the hashCode() method as well.